br/pkg/lightning/backend/local/duplicate.go

```diff
@@ -573,6 +591,117 @@ func (m *DupeDetector) RecordIndexConflictError(ctx context.Context, stream DupK
+	tErr, ok := errors.Cause(err).(*terror.Error)
+	if !ok {
+		return nil, nil, err
+	}
+	if len(tErr.Args()) != 2 {
+		return nil, nil, err
+	}
+	key, keyIsByte := tErr.Args()[0].([]byte)
+	value, valIsByte := tErr.Args()[1].([]byte)
+	if !keyIsByte || !valIsByte {
+		return nil, nil, err
+	}
+	return key, value, nil
+}
+
+// NewErrFoundConflictRecords generate an error ErrFoundDuplicateKeys
+// according to key and value.
+func NewErrFoundConflictRecords(key []byte, value []byte, tbl table.Table) error {
+	if tablecodec.IsRecordKey(key) {
+		// for data KV
+		sessionOpts := encode.SessionOptions{
+			SQLMode: mysql.ModeStrictAllTables,
+		}
+		encoder, err := kv.NewBaseKVEncoder(&encode.EncodingConfig{
+			Table:          tbl,
+			SessionOptions: sessionOpts,
+			Logger:         log.Logger{},
+		})
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		handle, err := tablecodec.DecodeRowKey(key)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		rowData, _, err := tables.DecodeRawRowData(encoder.SessionCtx,
+			tbl.Meta(), handle, tbl.Cols(), value)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		return errors.Trace(common.ErrFoundDataConflictRecords.FastGenByArgs(tbl.Meta().Name, handle.String(), rowData))
+	}
+
+	// for index KV
+	_, idxID, idxValues, err := tablecodec.DecodeIndexKey(key)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	idxInfo, err := GetIndexInfoByIndexID(tbl, idxID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	idxColLen := len(idxInfo.Columns)
+	indexName := fmt.Sprintf("%s.%s", tbl.Meta().Name.String(), idxInfo.Name.String())
+	colInfos := tables.BuildRowcodecColInfoForIndexColumns(idxInfo, tbl.Meta())
+	values, err := tablecodec.DecodeIndexKV(key, value, idxColLen, tablecodec.HandleNotNeeded, colInfos)
+	if err != nil {
+		tidblogutil.BgLogger().Warn("decode index key value failed", zap.String("index", indexName),
+			zap.String("key", hex.EncodeToString(key)), zap.String("value", hex.EncodeToString(value)), zap.Error(err))
+		return errors.Trace(common.ErrFoundIndexConflictRecords.FastGenByArgs(tbl.Meta().Name, key, value))
+	}
+	valueStr := make([]string, 0, idxColLen)
+	for i, val := range values[:idxColLen] {
+		d, err := tablecodec.DecodeColumnValue(val, colInfos[i].Ft, time.Local)
+		if err != nil {
+			tidblogutil.BgLogger().Warn("decode column value failed", zap.String("index", indexName),
+				zap.String("key", hex.EncodeToString(key)), zap.String("value", hex.EncodeToString(value)), zap.Error(err))
+			return errors.Trace(common.ErrFoundIndexConflictRecords.FastGenByArgs(tbl.Meta().Name, key, value))
+		}
+		str, err := d.ToString()
+		if err != nil {
+			str = string(val)
+		}
+		if types.IsBinaryStr(colInfos[i].Ft) || types.IsTypeBit(colInfos[i].Ft) {
+			str = tidbutil.FmtNonASCIIPrintableCharToHex(str)
+		}
+		valueStr = append(valueStr, str)
+	}
+
+	return errors.Trace(common.ErrFoundIndexConflictRecords.FastGenByArgs(tbl.Meta().Name, idxValues[0], valueStr[0]))
```

Comment:
I think here we only need to show any one value of the conflict key, to conform to the returned error of data KV, which also only shows one value of the conflict key. It may not be easy to change the code for data KV to show all the values of the conflict key.

Reply:
You can check what's the meaning of idxValues[0] or valueStr[0]. Maybe they are from a multiple column index, so all columns are needed to identify the row.